hazelcast/src/main/java/com/hazelcast/map/impl/MapIndexSynchronizer.java

```diff
@@ -0,0 +1,123 @@
+/*
+ * Copyright (c) 2008-2017, Hazelcast, Inc. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.hazelcast.map.impl;
+
+import com.hazelcast.internal.cluster.Versions;
+import com.hazelcast.internal.partition.InternalPartitionService;
+import com.hazelcast.internal.partition.impl.InternalPartitionServiceImpl;
+import com.hazelcast.logging.ILogger;
+import com.hazelcast.map.impl.operation.SynchronizeIndexesForPartitionTask;
+import com.hazelcast.query.impl.IndexInfo;
+import com.hazelcast.query.impl.MapIndexInfo;
+import com.hazelcast.spi.NodeEngine;
+import com.hazelcast.spi.OperationService;
+import com.hazelcast.spi.impl.operationservice.impl.OperationServiceImpl;
+import com.hazelcast.spi.partition.IPartitionService;
+import com.hazelcast.spi.serialization.SerializationService;
+import com.hazelcast.version.Version;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+
+// RU_COMPAT_V38
+
+/**
+ * This class is responsible for tracking cluster version changes and synchronized indexes if needed.
+ * The synchronization will happen only on 3.8 to 3.9 cluster version change.
+ *
+ * It is a last-chance anit-entropy that guards the situation where the index definitions arrive after the map data.
+ * In this case the MapContainer.indexesToAdd indexes will be send to each partition and the indexes will be populated.
+ *
+ * IMPORTANT: The synchronization applies to runtime partitioned indexes only.
+ * It is impossible to apply this fix for global indexes to the their global nature. It would require re-adding all data.
+ *
+ * @see com.hazelcast.map.impl.operation.SynchronizeIndexesForPartitionTask
+ * @see com.hazelcast.map.impl.operation.PostJoinMapOperation
+ * @see com.hazelcast.map.impl.operation.MapReplicationStateHolder
+ */
+class MapIndexSynchronizer {
```

Comment:
Just to be clear on a couple of points : 
- the index is filled in the migration commit phase (in MapMigrationAwareService#migrateIndex)?
- the job of the `MapIndexSynchronizer` is to cover the case where we add a new 3.9 node and it first receives from a 3.8 member the `MapReplicationOperation` and then later it receives the `PostJoinMapOperation`. This means that the index was added after the data replication. Is this right? Or does the synchronizer cover other cases?

If the second point is correct, then there are a couple of additional points:
- does that mean that the index is empty until it is filled with the synchronizer (on cluster version change)?
- not sure about this but when the index info list is being collected in the `PostJoinMapOperation` it does not take the `indexesToAdd` list into account. The cluster version change hasn't happened yet and the indexes are not created and populated.  The `PostJoinMapOperation` will then not send some indexInfos.

Reply:
a.) yes, it happens on finalize-migration for each partition
b.) the high-level problem was to move the HD index creation to partition-threads.
The MapIndexSynchronizer is just an temporary add-on to remove the race for HD indexes.
In 3.10 we can remove it, since the definitions will arrive in the MapReplicationOperation.
c.) yes, that is the case. without the synchronizer, the index will never be there.
The problem still holds for non-hd memory till 3.9+. See the comment in the MapIndexSynchronizer
```
* IMPORTANT: The synchronization applies to runtime partitioned indexes only.
 * It is impossible to apply this fix for global indexes due to the their global nature.
 * It would require re-adding all data, since you don't know if the data from this partition has been added
 * to the index just by checking if the index exists or not.
```

d.) That's true. I mean I assumed there will be at least a single migration in-between, but that's even better if I add it, so I did so.